test/editor/roundtrip-test.ts

```diff
@@ -0,0 +1,162 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { MatrixEvent } from 'matrix-js-sdk/src/matrix';
+
+import { parseEvent } from "../../src/editor/deserialize";
+import EditorModel from '../../src/editor/model';
+import DocumentOffset from '../../src/editor/offset';
+import { htmlSerializeIfNeeded, textSerialize } from '../../src/editor/serialize';
+import { createPartCreator } from "./mock";
+
+function htmlMessage(formattedBody: string, msgtype = "m.text") {
+    return {
+        getContent() {
+            return {
+                msgtype,
+                format: "org.matrix.custom.html",
+                formatted_body: formattedBody,
+            };
+        },
+    } as unknown as MatrixEvent;
+}
+
+async function md2html(markdown: string): Promise<string> {
+    const pc = createPartCreator();
+    const oldModel = new EditorModel([], pc, () => {});
+    await oldModel.update(
+        markdown,
+        "insertText",
+        new DocumentOffset(markdown.length, false),
+    );
+    return htmlSerializeIfNeeded(oldModel, { forceHTML: true });
+}
+
+function html2md(html: string): string {
+    const pc = createPartCreator();
+    const parts = parseEvent(htmlMessage(html), pc);
+    const newModel = new EditorModel(parts, pc);
+    return textSerialize(newModel);
+}
+
+async function roundTripMarkdown(markdown: string): Promise<string> {
+    return html2md(await md2html(markdown));
+}
+
+async function roundTripHtml(html: string): Promise<string> {
+    return await md2html(html2md(html));
+}
+
+describe('editor/roundtrip', function() {
+    describe('markdown messages should round-trip if they contain', function() {
+        test.each([
+            ["newlines", "hello\nworld"],
+            ["pills", "text message for @room"],
+            ["pills with interesting characters in mxid", "text message for @alice\\\\\\_\\]#>&:hs.example.com"],
+            ["styling", "**bold** and _emphasised_"],
+            ["bold within a word", "abso**fragging**lutely"],
+            ["escaped html", "a\\<foo>b"],
+            ["escaped markdown", "\\*\\*foo\\*\\* \\_bar\\_ \\[a\\](b)"],
+            ["escaped backslashes", "C:\\\\Program Files"],
+            ["code in backticks", "foo ->`x`"],
+            ["code blocks containing backticks", "```\nfoo ->`x`\nbar\n```"],
+            ["nested formatting", "a<del>b **c _d_ e** f</del>g"],
+            ["an ordered list", "A\n\n1. b\n2. c\n3. d\nE"],
+            ["an unordered list", "A\n\n- b\n- c\n- d\nE"],
+            ["f code block followed by text after a blank line", "```A\nfoo(bar).baz();\n\n3\n```\n\nB"],
```

Comment:
```suggestion
            ["code block followed by text after a blank line", "```A\nfoo(bar).baz();\n\n3\n```\n\nB"],
```